dm/relay/local_reader.go

```diff
@@ -567,66 +637,126 @@ func (r *BinlogReader) parseFile(
+		return false, false, nil
 	}
 
-	switchCh := make(chan SwitchPath, 1)
-	updatePathCh := make(chan string, 1)
-	updateErrCh := make(chan error, 1)
-	newCtx, cancel := context.WithCancel(ctx)
-	var wg sync.WaitGroup
-	defer func() {
-		cancel()
-		wg.Wait()
-	}()
+	return r.waitBinlogChanged(ctx, state)
+}
 
-	wg.Add(1)
-	go func(latestPos int64) {
-		defer wg.Done()
-		checker := relayLogFileChecker{
-			notifier:          r,
-			relayDir:          r.cfg.RelayDir,
-			currentUUID:       currentUUID,
-			latestRelayLogDir: relayLogDir,
-			latestFilePath:    fullPath,
-			latestFile:        relayLogFile,
-			beginOffset:       offset,
-			endOffset:         latestPos,
-		}
-		// TODO no need to be a goroutine now, maybe refactored when refactoring parseFile itself.
-		checker.relayLogUpdatedOrNewCreated(newCtx, updatePathCh, switchCh, updateErrCh)
-	}(latestPos)
+func (r *BinlogReader) waitBinlogChanged(ctx context.Context, state *binlogFileParseState) (needSwitch, needReParse bool, err error) {
+	active, relayOffset := r.relay.IsActive(r.currentUUID, state.relayLogFile)
+	if active && relayOffset > state.latestPos {
+		return false, true, nil
+	}
+	if !active {
+		meta := &LocalMeta{}
+		_, err := toml.DecodeFile(filepath.Join(state.relayLogDir, utils.MetaFilename), meta)
+		if err != nil {
+			return false, false, terror.Annotate(err, "decode relay meta toml file failed")
+		}
+		// current watched file size have no change means that no new writes have been made
+		// our relay meta file will be updated immediately after receive the rotate event,
+		// although we cannot ensure that the binlog filename in the meta is the next file after latestFile
+		// but if we return a different filename with latestFile, the outer logic (parseDirAsPossible)
+		// will find the right one
+		if meta.BinLogName != state.relayLogFile {
+			// we need check file size again, as the file may have been changed during our metafile check
+			cmp, err2 := fileSizeUpdated(state.fullPath, state.latestPos)
+			if err2 != nil {
+				return false, false, terror.Annotatef(err2, "latestFilePath=%s endOffset=%d", state.fullPath, state.latestPos)
+			}
+			switch {
+			case cmp < 0:
+				return false, false, terror.ErrRelayLogFileSizeSmaller.Generate(state.fullPath)
+			case cmp > 0:
+				return false, true, nil
+			default:
+				nextFilePath := filepath.Join(state.relayLogDir, meta.BinLogName)
+				log.L().Info("newer relay log file is already generated",
+					zap.String("now file path", state.fullPath),
+					zap.String("new file path", nextFilePath))
+				return false, false, nil
+			}
+		}
 
-	select {
-	case <-ctx.Done():
-		return false, false, 0, "", "", false, nil
-	case switchResp := <-switchCh:
-		// update new uuid
-		if err = r.updateUUIDs(); err != nil {
-			return false, false, 0, "", "", false, nil
+		// maybe UUID index file changed
+		switchPath, err := r.getSwitchPath()
```

Comment:
should we also updateUUIDs by the way? or merge them into one function

Reply:
not a frequent call, maybe later.